lib/irb.rb

```diff
@@ -598,25 +602,28 @@ def generate_prompt(opens, continue, line_offset)
     end
 
     def truncate_prompt_main(str) # :nodoc:
-      str = str.tr(CONTROL_CHARACTERS_PATTERN, ' ')
-      if str.size <= PROMPT_MAIN_TRUNCATE_LENGTH
-        str
-      else
-        str[0, PROMPT_MAIN_TRUNCATE_LENGTH - PROMPT_MAIN_TRUNCATE_OMISSION.size] + PROMPT_MAIN_TRUNCATE_OMISSION
+      if str.size > PROMPT_MAIN_TRUNCATE_LENGTH
+        str = str[0, PROMPT_MAIN_TRUNCATE_LENGTH - PROMPT_MAIN_TRUNCATE_OMISSION.size] + PROMPT_MAIN_TRUNCATE_OMISSION
       end
+      str.tr(CONTROL_CHARACTERS_PATTERN, ' ')
     end
 
     def format_prompt(format, ltype, indent, line_no) # :nodoc:
+      part_cache = @prompt_part_cache || {}
```

Comment:
Since `format_prompt` is only called in `generate_prompt`, is it possible to be called outside of `readmultiline` context? (perhaps in the `dyamic_prompt` callback?)

If it can be, let's also wrap the caller inside the wrapper mentioned above to provide caching?

Reply:
I see. Then let's at least add comments to explain why it could be nil, like

```suggestion
      # @prompt_part_cache could be nil in unit tests
      part_cache = @prompt_part_cache || {}
```